File: android/core/core-model/src/main/kotlin/com/fortunepocket/core/model/AstrologyChart.kt
package com.fortunepocket.core.model

import kotlinx.serialization.Serializable

@Serializable
enum class AstrologyPlanetID {
    SUN,
    MOON,
    MERCURY,
    VENUS,
    MARS,
    JUPITER,
    SATURN,
    URANUS,
    NEPTUNE,
    PLUTO;

    fun localizedName(isZh: Boolean): String = when (this) {
        SUN -> if (isZh) "太阳" else "Sun"
        MOON -> if (isZh) "月亮" else "Moon"
        MERCURY -> if (isZh) "水星" else "Mercury"
        VENUS -> if (isZh) "金星" else "Venus"
        MARS -> if (isZh) "火星" else "Mars"
        JUPITER -> if (isZh) "木星" else "Jupiter"
        SATURN -> if (isZh) "土星" else "Saturn"
        URANUS -> if (isZh) "天王星" else "Uranus"
        NEPTUNE -> if (isZh) "海王星" else "Neptune"
        PLUTO -> if (isZh) "冥王星" else "Pluto"
    }
}

@Serializable
enum class AstrologyAspectType(val angle: Double, val orb: Double) {
    CONJUNCTION(0.0, 8.0),
    SEXTILE(60.0, 4.5),
    SQUARE(90.0, 6.0),
    TRINE(120.0, 6.0),
    OPPOSITION(180.0, 8.0);

    fun localizedName(isZh: Boolean): String = when (this) {
        CONJUNCTION -> if (isZh) "合相" else "Conjunction"
        SEXTILE -> if (isZh) "六合" else "Sextile"
        SQUARE -> if (isZh) "刑克" else "Square"
        TRINE -> if (isZh) "拱相" else "Trine"
        OPPOSITION -> if (isZh) "冲相" else "Opposition"
    }
}

@Serializable
data class AstrologyPlanetPlacement(
    val planetId: AstrologyPlanetID,
    val longitude: Double,
    val signId: String,
    val signNameZh: String,
    val signNameEn: String,
    val house: Int? = null,
    val isRetrograde: Boolean = false
) {
    fun localizedSignName(isZh: Boolean): String = if (isZh) signNameZh else signNameEn
    val degreeInSign: Double get() = ((longitude % 30.0) + 30.0) % 30.0
}

@Serializable
data class AstrologyAspect(
    val firstPlanetId: AstrologyPlanetID,
    val secondPlanetId: AstrologyPlanetID,
    val type: AstrologyAspectType,
    val orbDegrees: Double
)

@Serializable
data class AstrologyElementBalance(
    val fire: Int,
    val earth: Int,
    val air: Int,
    val water: Int
) {
    val dominantElementKey: String
        get() = listOf(
            "fire" to fire,
            "earth" to earth,
            "air" to air,
            "water" to water
        ).maxWithOrNull(compareBy<Pair<String, Int>> { it.second }.thenByDescending { it.first })?.first ?: "fire"

    fun localizedDominantElement(isZh: Boolean): String = when (dominantElementKey) {
        "fire" -> if (isZh) "火象" else "Fire"
        "earth" -> if (isZh) "土象" else "Earth"
        "air" -> if (isZh) "风象" else "Air"
        "water" -> if (isZh) "水象" else "Water"
        else -> if (isZh) "火象" else "Fire"
    }
}

@Serializable
data class AstrologyHouseFocus(
    val house: Int,
    val titleZh: String,
    val titleEn: String,
    val summaryZh: String,
    val summaryEn: String
) {
    fun localizedTitle(isZh: Boolean): String = if (isZh) titleZh else titleEn
    fun localizedSummary(isZh: Boolean): String = if (isZh) summaryZh else summaryEn
}

data class HoroscopeReading(
    val sign: ZodiacSign,
    val moonSign: ZodiacSign,
    val risingSign: ZodiacSign?,
    val birthDateText: String,
    val birthTimeText: String,
    val birthCityName: String,
    val timeZoneId: String,
    val chartSummary: String,
    val chartSignature: String,
    val overall: String,
    val love: String,
    val career: String,
    val wealth: String,
    val social: String,
    val advice: String,
    val luckyColor: String,
    val luckyNumber: Int,
    val planetPlacements: List<AstrologyPlanetPlacement>,
    val majorAspects: List<AstrologyAspect>,
    val elementBalance: AstrologyElementBalance,
    val houseFocus: List<AstrologyHouseFocus>
)

File: ios/PocketOracle/Models/AstrologyChart.swift
import Foundation

enum AstrologyPlanetID: String, Codable, CaseIterable, Identifiable {
    case sun
    case moon
    case mercury
    case venus
    case mars
    case jupiter
    case saturn
    case uranus
    case neptune
    case pluto

    var id: String { rawValue }

    var localizedName: String {
        switch self {
        case .sun: return AppLanguageOption.isChinese ? "太阳" : "Sun"
        case .moon: return AppLanguageOption.isChinese ? "月亮" : "Moon"
        case .mercury: return AppLanguageOption.isChinese ? "水星" : "Mercury"
        case .venus: return AppLanguageOption.isChinese ? "金星" : "Venus"
        case .mars: return AppLanguageOption.isChinese ? "火星" : "Mars"
        case .jupiter: return AppLanguageOption.isChinese ? "木星" : "Jupiter"
        case .saturn: return AppLanguageOption.isChinese ? "土星" : "Saturn"
        case .uranus: return AppLanguageOption.isChinese ? "天王星" : "Uranus"
        case .neptune: return AppLanguageOption.isChinese ? "海王星" : "Neptune"
        case .pluto: return AppLanguageOption.isChinese ? "冥王星" : "Pluto"
        }
    }
}

enum AstrologyAspectType: String, Codable, CaseIterable, Identifiable {
    case conjunction
    case sextile
    case square
    case trine
    case opposition

    var id: String { rawValue }

    var angle: Double {
        switch self {
        case .conjunction: return 0
        case .sextile: return 60
        case .square: return 90
        case .trine: return 120
        case .opposition: return 180
        }
    }

    var orb: Double {
        switch self {
        case .conjunction: return 8
        case .sextile: return 4.5
        case .square: return 6
        case .trine: return 6
        case .opposition: return 8
        }
    }

    var localizedName: String {
        switch self {
        case .conjunction: return AppLanguageOption.isChinese ? "合相" : "Conjunction"
        case .sextile: return AppLanguageOption.isChinese ? "六合" : "Sextile"
        case .square: return AppLanguageOption.isChinese ? "刑克" : "Square"
        case .trine: return AppLanguageOption.isChinese ? "拱相" : "Trine"
        case .opposition: return AppLanguageOption.isChinese ? "冲相" : "Opposition"
        }
    }
}

struct AstrologyPlanetPlacement: Codable, Identifiable {
    let planetId: AstrologyPlanetID
    let longitude: Double
    let signId: String
    let signNameZh: String
    let signNameEn: String
    let house: Int?
    let isRetrograde: Bool

    var id: String { planetId.rawValue }

    var localizedPlanetName: String { planetId.localizedName }
    var localizedSignName: String { AppLanguageOption.isChinese ? signNameZh : signNameEn }
    var degreeInSign: Double { longitude.truncatingRemainder(dividingBy: 30) }
}

struct AstrologyAspect: Codable, Identifiable {
    let firstPlanetId: AstrologyPlanetID
    let secondPlanetId: AstrologyPlanetID
    let type: AstrologyAspectType
    let orbDegrees: Double

    var id: String { "\(firstPlanetId.rawValue)-\(secondPlanetId.rawValue)-\(type.rawValue)" }
}

struct AstrologyElementBalance: Codable {
    let fire: Int
    let earth: Int
    let air: Int
    let water: Int

    var dominantElementKey: String {
        [
            ("fire", fire),
            ("earth", earth),
            ("air", air),
            ("water", water)
        ]
        .max { lhs, rhs in
            if lhs.1 == rhs.1 {
                return lhs.0 > rhs.0
            }
            return lhs.1 < rhs.1
        }?
        .0 ?? "fire"
    }

    var localizedDominantElement: String {
        switch dominantElementKey {
        case "fire": return AppLanguageOption.isChinese ? "火象" : "Fire"
        case "earth": return AppLanguageOption.isChinese ? "土象" : "Earth"
        case "air": return AppLanguageOption.isChinese ? "风象" : "Air"
        case "water": return AppLanguageOption.isChinese ? "水象" : "Water"
        default: return AppLanguageOption.isChinese ? "火象" : "Fire"
        }
    }
}

struct AstrologyHouseFocus: Codable, Identifiable {
    let house: Int
    let titleZh: String
    let titleEn: String
    let summaryZh: String
    let summaryEn: String

    var id: Int { house }
    var localizedTitle: String { AppLanguageOption.isChinese ? titleZh : titleEn }
    var localizedSummary: String { AppLanguageOption.isChinese ? summaryZh : summaryEn }
}

struct HoroscopeReading {
    let sign: ZodiacSign
    let moonSign: ZodiacSign
    let risingSign: ZodiacSign?
    let birthDateText: String
    let birthTimeText: String
    let birthCityName: String
    let timeZoneId: String
    let chartSummary: String
    let chartSignature: String
    let personalityCore: String
    let relationshipPattern: String
    let strengths: String
    let growthEdge: String
    let currentTheme: String
    let overall: String
    let love: String
    let career: String
    let wealth: String
    let social: String
    let advice: String
    let luckyColor: String
    let luckyNumber: Int
    let planetPlacements: [AstrologyPlanetPlacement]
    let majorAspects: [AstrologyAspect]
    let elementBalance: AstrologyElementBalance
    let houseFocus: [AstrologyHouseFocus]
}
